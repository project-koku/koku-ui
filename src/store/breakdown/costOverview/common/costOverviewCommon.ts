import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum CostOverviewWidgetType {
  cluster = 'cluster', // This type displays clusters associated with a project
  cost = 'cost', // This type displays a cost breakdown as a pie chart
  cpuUsage = 'cpuUsage', // This type displays cpu usage as a bullet chart
  memoryUsage = 'memoryUsage', // This type displays memory usage as a bullet chart
  reportSummary = 'summary', // This type displays a cost report summary
  volumeUsage = 'volumeUsage', // This type displays volume usage as a bullet chart
}

export interface CostOverviewWidget {
  chartName?: string; // Will be the prefix for ids within the chart
  id: number;
  cluster?: {
    reportGroupBy: string; // Report group_by
    showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
  };
  usage?: {
    showCapacityOnGroupBy?: string[]; // Show capacity when group_by is matched
  };
  reportSummary?: {
    reportGroupBy: string; // Report group_by
    showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
    usePlaceholder?: boolean; // Use placeholder to keep card placement when widget is not shown
  };
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  type: CostOverviewWidgetType;
}
