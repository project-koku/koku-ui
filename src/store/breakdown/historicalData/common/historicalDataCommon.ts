import type { ReportPathsType, ReportType } from 'api/reports/report';

export const enum HistoricalDataWidgetType {
  cost = 'cost', // This type displays historical cost chart
  network = 'network', // This type displays historical network chart
  trend = 'trend', // This type displays historical trend chart
  usage = 'usage', // This type displays historical usage chart
  volume = 'volume', // This type displays historical volume chart
}

export interface HistoricalDataWidget {
  chartName: string; // Will be the prefix for ids within the chart
  id: number;
  reportPathsType: ReportPathsType; // Report URL path
  reportType: ReportType; // Report type; cost, storage, etc.
  showLimit?: boolean; // Show chart legend for limit data
  showRequest?: boolean; // Show chart legend for request data
  showWidgetOnGroupBy?: string[]; // Show widget when group_by is matched
  type: HistoricalDataWidgetType;
}
