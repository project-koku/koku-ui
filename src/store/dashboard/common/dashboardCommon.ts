import { MessageDescriptor } from '@formatjs/intl/src/types';
import { ForecastPathsType, ForecastType } from 'api/forecasts/forecast';
import { ReportPathsType, ReportType } from 'api/reports/report';

// eslint-disable-next-line no-shadow
export const enum DashboardChartType {
  cost = 'cost', // // This displays cumulative cost compared to infrastructure cost
  dailyCost = 'dailyCost', // This displays cumulative and daily cost compared to infrastructure cost
  dailyTrend = 'dailyTrend', // This displays cumulative and daily cost
  trend = 'trend', // This displays cumulative cost only
  usage = 'usage', // This displays daily usage and requests
}

export interface ValueFormatOptions {
  fractionDigits?: number;
}

export interface DashboardWidget<T> {
  availableTabs?: T[];
  chartType?: DashboardChartType;
  currentTab?: T;
  details: {
    adjustContainerHeight?: boolean; // Adjust chart container height for responsiveness
    costKey?: MessageDescriptor; // i18n key
    formatOptions: ValueFormatOptions;
    requestFormatOptions?: {
      fractionDigits?: number;
    };
    requestKey?: MessageDescriptor;
    showHorizontal?: boolean; // Show horizontal layout
    showTooltip?: boolean; // Show cost tooltip
    showUnits?: boolean; // Show units
    showUsageFirst?: boolean; // Show usage before cost
    showUsageLegendLabel?: boolean;
    units?: string; // Override units shown as workaround for missing Azure API units
    usageFormatOptions?: ValueFormatOptions;
    usageKey?: MessageDescriptor; // i18n key
    viewAllPath?: string; // View all link to details page
  };
  filter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  forecastPathsType?: ForecastPathsType;
  forecastType?: ForecastType;
  id: number;
  reportPathsType: ReportPathsType;
  reportType: ReportType;
  /** i18n key for the title. passed { startDate, endDate, month, time } */
  titleKey: MessageDescriptor;
  tabsFilter?: {
    limit?: number;
    service?: string;
    service_name?: string;
  };
  trend: {
    computedForecastItem?: string; // The computed forecast item to use in charts.
    computedForecastInfrastructureItem?: string; // The computed forecast infrastructure item to use in charts.
    computedReportItem: string; // The computed report item to use in charts, summary, etc.
    computedReportItemValue: string; // The computed report value (e.g., raw, markup, total, or usage)
    dailyTitleKey?: MessageDescriptor;
    showInfrastructureLabel?: boolean; // Trend chart legend items show "Infrastructure cost" instead of "cost"
    showSupplementaryLabel?: boolean; // Trend chart legend items show "Supplementary cost" instead of "cost"
    titleKey: MessageDescriptor;
    type: number;
    formatOptions: ValueFormatOptions;
  };
  topItems?: {
    formatOptions: any;
  };
}
