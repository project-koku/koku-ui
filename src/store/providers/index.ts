import {
  awsProvidersQuery,
  azureProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ociProvidersQuery,
  ocpProvidersQuery,
  providersQuery,
} from 'store/providers/providersCommon';

import * as providersActions from './providersActions';
import { stateKey as providersStateKey } from './providersCommon';
import { ProvidersAction, providersReducer, ProvidersState } from './providersReducer';
import * as providersSelectors from './providersSelectors';

export {
  awsProvidersQuery,
  azureProvidersQuery,
  ociProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ocpProvidersQuery,
  ProvidersAction,
  providersActions,
  providersQuery,
  providersReducer,
  providersSelectors,
  ProvidersState,
  providersStateKey,
};
