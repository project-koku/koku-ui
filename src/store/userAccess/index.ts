import {
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessQuery,
} from 'store/userAccess/userAccessCommon';

import * as userAccessActions from './userAccessActions';
import { stateKey as userAccessStateKey } from './userAccessCommon';
import type { UserAccessAction, UserAccessState } from './userAccessReducer';
import { userAccessReducer } from './userAccessReducer';
import * as userAccessSelectors from './userAccessSelectors';

export {
  userAccessQuery,
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessActions,
  userAccessReducer,
  userAccessSelectors,
  userAccessStateKey,
};
export type { UserAccessAction, UserAccessState };
