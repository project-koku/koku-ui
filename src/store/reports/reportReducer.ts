import { Report } from 'api/reports/report';
import { AxiosError } from 'axios';
import { FetchStatus } from 'store/common';
import { resetState } from 'store/ui/uiActions';
import { ActionType, getType } from 'typesafe-actions';

import { fetchReportFailure, fetchReportRequest, fetchReportSuccess } from './reportActions';

export interface CachedReport extends Report {
  timeRequested: number;
}

export type ReportState = Readonly<{
  byId: Map<string, CachedReport>;
  fetchStatus: Map<string, FetchStatus>;
  errors: Map<string, AxiosError>;
}>;

const defaultState: ReportState = {
  byId: new Map(),
  fetchStatus: new Map(),
  errors: new Map(),
};

export type ReportAction = ActionType<
  typeof fetchReportFailure | typeof fetchReportRequest | typeof fetchReportSuccess | typeof resetState
>;

export function reportReducer(state = defaultState, action: ReportAction): ReportState {
  switch (action.type) {
    case getType(resetState):
      state = defaultState;
      return state;

    case getType(fetchReportRequest):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(action.payload.reportId, FetchStatus.inProgress),
      };

    case getType(fetchReportSuccess):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(action.meta.reportId, FetchStatus.complete),
        byId: new Map(state.byId).set(action.meta.reportId, {
          ...action.payload,
          timeRequested: Date.now(),
        }),
        errors: new Map(state.errors).set(action.meta.reportId, null),
      };

    case getType(fetchReportFailure):
      return {
        ...state,
        fetchStatus: new Map(state.fetchStatus).set(action.meta.reportId, FetchStatus.complete),
        errors: new Map(state.errors).set(action.meta.reportId, action.payload),
      };
    default:
      return state;
  }
}
