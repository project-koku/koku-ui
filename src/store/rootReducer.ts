import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import { combineReducers } from 'redux';
import { awsCostOverviewReducer, awsCostOverviewStateKey } from 'store/breakdown/costOverview/awsCostOverview';
import { azureCostOverviewReducer, azureCostOverviewStateKey } from 'store/breakdown/costOverview/azureCostOverview';
import { gcpCostOverviewReducer, gcpCostOverviewStateKey } from 'store/breakdown/costOverview/gcpCostOverview';
import { ibmCostOverviewReducer, ibmCostOverviewStateKey } from 'store/breakdown/costOverview/ibmCostOverview';
import { ocpCostOverviewReducer, ocpCostOverviewStateKey } from 'store/breakdown/costOverview/ocpCostOverview';
import { awsHistoricalDataReducer, awsHistoricalDataStateKey } from 'store/breakdown/historicalData/awsHistoricalData';
import {
  azureHistoricalDataReducer,
  azureHistoricalDataStateKey,
} from 'store/breakdown/historicalData/azureHistoricalData';
import { gcpHistoricalDataReducer, gcpHistoricalDataStateKey } from 'store/breakdown/historicalData/gcpHistoricalData';
import { ibmHistoricalDataReducer, ibmHistoricalDataStateKey } from 'store/breakdown/historicalData/ibmHistoricalData';
import { ocpHistoricalDataReducer, ocpHistoricalDataStateKey } from 'store/breakdown/historicalData/ocpHistoricalData';
import { costModelsReducer, costModelsStateKey } from 'store/costModels';
import { awsDashboardReducer, awsDashboardStateKey } from 'store/dashboard/awsDashboard';
import { awsOcpDashboardReducer, awsOcpDashboardStateKey } from 'store/dashboard/awsOcpDashboard';
import { azureDashboardReducer, azureDashboardStateKey } from 'store/dashboard/azureDashboard';
import { azureOcpDashboardReducer, azureOcpDashboardStateKey } from 'store/dashboard/azureOcpDashboard';
import { gcpDashboardReducer, gcpDashboardStateKey } from 'store/dashboard/gcpDashboard';
import { gcpOcpDashboardReducer, gcpOcpDashboardStateKey } from 'store/dashboard/gcpOcpDashboard';
import { ibmDashboardReducer, ibmDashboardStateKey } from 'store/dashboard/ibmDashboard';
import { ocpCloudDashboardReducer, ocpCloudDashboardStateKey } from 'store/dashboard/ocpCloudDashboard';
import { ocpDashboardReducer, ocpDashboardStateKey } from 'store/dashboard/ocpDashboard';
import {
  ocpInfrastructureDashboardReducer,
  ocpInfrastructureDashboardStateKey,
} from 'store/dashboard/ocpInfrastructureDashboard';
import {
  ocpSupplementaryDashboardReducer,
  ocpSupplementaryDashboardStateKey,
} from 'store/dashboard/ocpSupplementaryDashboard';
import { ocpUsageDashboardReducer, ocpUsageDashboardStateKey } from 'store/dashboard/ocpUsageDashboard';
import { exportReducer, exportStateKey } from 'store/exports';
import { forecastReducer, forecastStateKey } from 'store/forecasts';
import { orgReducer, orgStateKey } from 'store/orgs';
import { priceListReducer, priceListStateKey } from 'store/priceList';
import { reportReducer, reportStateKey } from 'store/reports';
import { resourceReducer, resourceStateKey } from 'store/resources';
import { sourcesReducer, sourcesStateKey } from 'store/sourceSettings';
import { tagReducer, tagStateKey } from 'store/tags';
import { StateType } from 'typesafe-actions';

import { metricsReducer, metricsStateKey } from './metrics';
import { providersReducer, providersStateKey } from './providers';
import { rbacReducer, rbacStateKey } from './rbac';
import { uiReducer, uiStateKey } from './ui';
import { userAccessReducer, userAccessStateKey } from './userAccess';

export type RootState = StateType<typeof rootReducer>;

export const rootReducer = combineReducers({
  [awsOcpDashboardStateKey]: awsOcpDashboardReducer,
  [awsCostOverviewStateKey]: awsCostOverviewReducer,
  [awsDashboardStateKey]: awsDashboardReducer,
  [awsHistoricalDataStateKey]: awsHistoricalDataReducer,
  [azureOcpDashboardStateKey]: azureOcpDashboardReducer,
  [azureCostOverviewStateKey]: azureCostOverviewReducer,
  [azureDashboardStateKey]: azureDashboardReducer,
  [azureHistoricalDataStateKey]: azureHistoricalDataReducer,
  [costModelsStateKey]: costModelsReducer,
  [exportStateKey]: exportReducer,
  [gcpCostOverviewStateKey]: gcpCostOverviewReducer,
  [gcpDashboardStateKey]: gcpDashboardReducer,
  [gcpOcpDashboardStateKey]: gcpOcpDashboardReducer,
  [gcpHistoricalDataStateKey]: gcpHistoricalDataReducer,
  [ibmCostOverviewStateKey]: ibmCostOverviewReducer,
  [ibmDashboardStateKey]: ibmDashboardReducer,
  [ibmHistoricalDataStateKey]: ibmHistoricalDataReducer,
  [metricsStateKey]: metricsReducer,
  [ocpCostOverviewStateKey]: ocpCostOverviewReducer,
  [ocpDashboardStateKey]: ocpDashboardReducer,
  [ocpCloudDashboardStateKey]: ocpCloudDashboardReducer,
  [ocpHistoricalDataStateKey]: ocpHistoricalDataReducer,
  [ocpInfrastructureDashboardStateKey]: ocpInfrastructureDashboardReducer,
  [ocpSupplementaryDashboardStateKey]: ocpSupplementaryDashboardReducer,
  [ocpUsageDashboardStateKey]: ocpUsageDashboardReducer,
  [orgStateKey]: orgReducer,
  [priceListStateKey]: priceListReducer,
  [providersStateKey]: providersReducer,
  [rbacStateKey]: rbacReducer,
  [reportStateKey]: reportReducer,
  [forecastStateKey]: forecastReducer,
  [resourceStateKey]: resourceReducer,
  [sourcesStateKey]: sourcesReducer,
  [tagStateKey]: tagReducer,
  [uiStateKey]: uiReducer,
  [userAccessStateKey]: userAccessReducer,
  notifications,
});
