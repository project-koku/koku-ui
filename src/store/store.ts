import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';

import { rootReducer, RootState } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
