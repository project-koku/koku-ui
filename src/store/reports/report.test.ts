jest.mock('api/reports/reportUtils');

import { waitFor } from '@testing-library/react';
import type { Report } from 'api/reports/report';
import { ReportPathsType, ReportType } from 'api/reports/report';
import { runReport } from 'api/reports/reportUtils';
import { FetchStatus } from 'store/common';
import { createMockStoreCreator } from 'store/mockStore';

import * as actions from './reportActions';
import { reportStateKey } from './reportCommon';
import { reportReducer } from './reportReducer';
import * as selectors from './reportSelectors';

const createReportsStore = createMockStoreCreator({
  [reportStateKey]: reportReducer,
});

const runReportMock = runReport as jest.Mock;

const mockReport: Report = {
  data: [],
  total: {
    value: 100,
    units: 'USD',
  },
} as any;

const reportType = ReportType.cost;
const reportPathsType = ReportPathsType.aws;
const reportQueryString = 'reportQueryString';

runReportMock.mockResolvedValue({ data: mockReport });
global.Date.now = jest.fn(() => 12345);

jest.spyOn(actions, 'fetchReport');
jest.spyOn(selectors, 'selectReportFetchStatus');

test('default state', () => {
  const store = createReportsStore();
  expect(selectors.selectReportState(store.getState())).toMatchSnapshot();
});

test('fetch report success', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  expect(runReportMock).toHaveBeenCalled();
  expect(selectors.selectReportFetchStatus(store.getState(), reportPathsType, reportType, reportQueryString)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectReportFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectReportFetchStatus(finishedState, reportPathsType, reportType, reportQueryString)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectReportError(finishedState, reportPathsType, reportType, reportQueryString)).toBe(null);
});

test('fetch report failure', async () => {
  const store = createReportsStore();
  const error = Symbol('report error');
  runReportMock.mockRejectedValueOnce(error);
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  expect(runReport).toHaveBeenCalled();
  expect(selectors.selectReportFetchStatus(store.getState(), reportPathsType, reportType, reportQueryString)).toBe(
    FetchStatus.inProgress
  );
  await waitFor(() => expect(selectors.selectReportFetchStatus).toHaveBeenCalled());
  const finishedState = store.getState();
  expect(selectors.selectReportFetchStatus(finishedState, reportPathsType, reportType, reportQueryString)).toBe(
    FetchStatus.complete
  );
  expect(selectors.selectReportError(finishedState, reportPathsType, reportType, reportQueryString)).toBe(error);
});

test('does not fetch report if the request is in progress', () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  expect(runReport).toHaveBeenCalledTimes(1);
});

test('report is not refetched if it has not expired', async () => {
  const store = createReportsStore();
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  await waitFor(() => expect(actions.fetchReport).toHaveBeenCalled());
  store.dispatch(actions.fetchReport(reportPathsType, reportType, reportQueryString));
  expect(runReport).toHaveBeenCalledTimes(1);
});
