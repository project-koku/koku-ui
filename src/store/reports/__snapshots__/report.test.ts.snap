// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default state 1`] = `
{
  "byId": Map {},
  "errors": Map {},
  "fetchStatus": Map {},
}
`;
